src/python/zquantum/optimizers/layerwise_ansatz_optimizer.py

```diff
@@ -1,96 +1,175 @@
+            from zquantum.core.estimation import (
+                estimate_expectation_values_by_averaging,
+                allocate_shots_uniformly
+            )
+            from zquantum.core.cost_function import (
+                substitution_based_estimation_tasks_factory,
+                create_cost_function,
+            )
+
+            cost_hamiltonian = ...
+            ansatz = ...
+
+            estimation_preprocessors = [partial(allocate_shots_uniformly, number_of_shots=1000)]
+            estimation_tasks_factory_generator = partial(
+                substitution_based_estimation_tasks_factory,
+                target_operator=cost_hamiltonian,
+                estimation_preprocessors=estimation_preprocessors
+            )
+            cost_function_factory = partial(
+                create_cost_function,
+                backend=QuantumBackend,
+                estimation_method=estimate_expectation_values_by_averaging,
+                parameter_preprocessors=None,
+            )
+
+            initial_params = np.array([0.42, 4.2])
+            inner_optimizer = ...
+
+            optimizer = LayerwiseAnsatzOptimizer(
+                ansatz,
+                inner_optimizer,
+                estimation_tasks_factory_generator,
+                cost_function_factory,
+                min_layer = ...,
+                max_layer = ...,
+            )
+
+            opt_result = optimizer.minimize(initial_params)
         """
+        # TODO which should be private, which should not be private?
         assert 0 <= min_layer <= max_layer
         assert n_layers_per_iteration > 0
-        self.inner_optimizer = inner_optimizer
+        self._ansatz = ansatz
+        self._inner_optimizer = inner_optimizer
         self.min_layer = min_layer
         self.max_layer = max_layer
         self.n_layers_per_iteration = n_layers_per_iteration
-        if parameters_initializer is None:
-            self.parameters_initializer = partial(np.random.uniform, -np.pi, np.pi)
-        else:
-            self.parameters_initializer = parameters_initializer
-        self.recorder = recorder
+        self.parameters_initializer = parameters_initializer
+        self._recorder = recorder
 
-    def minimize(
+    def _minimize(
         self,
-        cost_function: Union[CallableWithGradient, Callable],
+        cost_function_factory: Callable[[Ansatz], CostFunction],
         initial_params: np.ndarray,
         keep_history: bool = False,
     ) -> OptimizeResult:
         """
         Finds the parameters which minimize given cost function, by trying all the parameters from the grid.
 
-        Note:
-            Cost function needs to have `ansatz` property
-
         Args:
-            cost_function(zquantum.core.interfaces.cost_function.CostFunction): object representing cost function we want to minimize
-            inital_params (np.ndarray): initial parameters for the cost function
+            # TODO
```

Comment:
Should we address this TODO? It seems that only `keep_history` and the description of the return value are missing from the docstring.